### PR TITLE
Introduce `dedupeFragments` config flag for dealing with duplicate fragments 

### PR DESCRIPTION
**Pull request by gilgardosh**
Related #4684

### CODE REVIEW

**Comment by dotansimha** [comment]
Thoughts on that:
1. I think while generating fragments `DocuemntNode`, we can flatten and remove ALL other fragments' definitions. So each fragment will be standalone, with just spreads for other fragments.
2. Then, in each query `DocumentNode` that we generate, we can add all fragments in the query, and deduplicate it, based on the map of fragments.

### Today

```ts
const fragmentB = `fragment B on B { ... }`;
const fragmentA = `fragment A on A { field { ...B } } ${fragmentB}`;
const query = `query { ...B, ...A } ${fragmentA} ${fragmentB}`
```

What I suggest, is to do:

```ts
const fragmentB = `fragment B on B { ... }`;
const fragmentA = `fragment A on A { field { ...B } }`; // No template literals here
const query = `query { ...B, ...A } ${fragmentA} ${fragmentB}` // Only here
```

File: packages/plugins/other/visitor-plugin-common/src/base-visitor.ts [hunk 1]
```diff
@@ -186,6 +187,10 @@ export interface RawConfig {
    * @ignore
    */
   globalNamespace?: boolean;
+  /**
```
Comment (dotansimha) [inline packages/plugins/other/visitor-plugin-common/src/base-visitor.ts:190]: Why `ignore`?

File: packages/plugins/other/visitor-plugin-common/src/base-visitor.ts [hunk 2]
```diff
@@ -186,6 +187,10 @@ export interface RawConfig {
...
+   * @ignore
```
Comment (dotansimha) [inline packages/plugins/other/visitor-plugin-common/src/base-visitor.ts:191]: Why `ignore`? I think we need to document it :)

Comment (dotansimha) [reply packages/plugins/other/visitor-plugin-common/src/base-visitor.ts:191]: You can define it here, and then you don't need to define it in the config of typed-document-node. This is relevant for all of the plugins that inherits.

File: packages/plugins/typescript/typed-document-node/tests/typed-document-node.spec.ts [hunk 3]
```diff
@@ -7,4 +8,54 @@ describe('TypeDocumentNode', () => {
...
+
+      type Job {
+        id: ID!
+        recruiterName: String!
+        title: String!
+      }
+    `);
+
+    const ast = parse(/* GraphQL */ `
+      query GetJobs {
```
Comment (dotansimha) [inline packages/plugins/typescript/typed-document-node/tests/typed-document-node.spec.ts:30]: Please add an additional test where the fragment is loaded also by the query (like this one, but we `...JobSimpleRecruiterData` as well)

**Review by dotansimha — APPROVED** [review]

File: packages/plugins/other/visitor-plugin-common/src/base-visitor.ts (hunk 2 above)
Comment (gilgardosh) [reply packages/plugins/other/visitor-plugin-common/src/base-visitor.ts:191]: It's documented [here](https://github.com/dotansimha/graphql-code-generator/blob/c47bcba43904ca6abd1da4209e07011ddd7c08d2/packages/plugins/typescript/typed-document-node/src/config.ts#L26), do you think it is relevant to activate it from the base visitor as well?

**Review by dotansimha — APPROVED** [review]

**Pull request merged**